From e134b11db46ed22c7a410c97265a8b886719b573 Mon Sep 17 00:00:00 2001
From: OpenEmbedded <mario.schuknecht@dresearch-fe.de>
Date: Mon, 5 Dec 2022 08:09:20 +0100
Subject: [PATCH] fec_main: Add dma_rmb() HYP-25259

Avoid fec error:
kernel: fec 2188000.ethernet eth0: rcv is not +last

The bit test of 'status' variable does not seem to work reliably. Valid
frames are rejected. If a dma_rmb() is inserted, the error can no longer
be observed.

Signed-off-by: Mario Schuknecht <mario.schuknecht@dresearch-fe.de>
---
 drivers/net/ethernet/freescale/fec_main.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/net/ethernet/freescale/fec_main.c b/drivers/net/ethernet/freescale/fec_main.c
index 111056dd6887..d0858bcfa2d1 100644
--- a/drivers/net/ethernet/freescale/fec_main.c
+++ b/drivers/net/ethernet/freescale/fec_main.c
@@ -1544,6 +1544,7 @@ static int fec_rxq(struct net_device *ndev, struct fec_enet_priv_rx_q *rxq,
 	 */
 	bdp = rxq->bd.cur;
 
+	dma_rmb();
 	while (pkt_received < budget) {
 		status = fec16_to_cpu(READ_ONCE(bdp->cbd_sc));
 		if (status & BD_ENET_RX_EMPTY)
-- 
2.35.3

